### PR TITLE
Signature conversions command-line flag

### PR DESCRIPTION
**Pull request by dakenf**
A simple change to add linker flag like `-sSIGNATURE_CONVERSIONS=OrtRun:_pppp`

This is going to be needed for any project that targets both 32 and 64bit as a lot of BigInt conversions and checks would be required in the client code

### CODE REVIEW

File: src/settings.js [hunk 1]
```diff
@@ -2060,6 +2060,11 @@ var RUNTIME_DEBUG = false;
 // library symbol.
 var LEGACY_RUNTIME = false;
 
+// User-defined functions to wrap with signature conversion
+// example use -sSIGNATURE_CONVERSIONS=['someFunction:_p']
+// [link]
+var SIGNATURE_CONVERSIONS = [];
```
Comment (sbc100) [inline src/settings.js:2066]: I've trying to think of some way that we can embed this in the source code itself.

Something like `__attribute__(emcc_sig("_p"))`.  

Perhaps we could add this setting in the interim?

File: src/settings.js [hunk 2]
```diff
@@ -2060,6 +2060,11 @@ var RUNTIME_DEBUG = false;
 // library symbol.
 var LEGACY_RUNTIME = false;
 
+// User-defined functions to wrap with signature conversion
+// example use -sSIGNATURE_CONVERSIONS=['someFunction:_p']
```
Comment (sbc100) [inline src/settings.js:2064]: You don't need the square brackets here, or the quotes.

File: src/settings.js [hunk 3]
```diff
@@ -2060,6 +2060,11 @@ var RUNTIME_DEBUG = false;
 // library symbol.
 var LEGACY_RUNTIME = false;
 
+// User-defined functions to wrap with signature conversion
```
Comment (sbc100) [inline src/settings.js:2063]: This comment should mention that this is for documenting native functions that take or return pointer arguments and that it only effects wasm64.

File: emscripten.py [hunk 4]
```diff
@@ -905,6 +905,11 @@ def create_pointer_conversion_wrappers(metadata):
     'emscripten_wasm_worker_initialize': '_p_',
   }
 
+  if settings.SIGNATURE_CONVERSIONS:
+      for function in settings.SIGNATURE_CONVERSIONS:
```
Comment (sbc100) [inline emscripten.py:909]: You don't need the outer if here.

File: src/settings.js (hunk 1 above)
Comment (dakenf) [reply src/settings.js:2066]: Do you want to somehow wrap them automatically with that? Will check what I can do

**Comment by dakenf** [comment]
I think this one is safe as the only test failed is "test_pthread_proxying_refcount" that sometimes fails randomly

File: emscripten.py [hunk 5]
```diff
@@ -883,6 +883,10 @@ def create_pointer_conversion_wrappers(metadata):
     '__get_exception_message': '_ppp',
   }
 
+  for function in settings.SIGNATURE_CONVERSIONS:
+    sig = function.split(':')
```
Comment (sbc100) [inline emscripten.py:887]: How about:
```
  sym, sig = function.split(':')` 
  mapping[sym] = sig
```

**Review by sbc100 — APPROVED** [review]
lgtm with final comment

File: emscripten.py (hunk 5 above)
Comment (dakenf) [reply emscripten.py]: Here you go

**Comment by dakenf** [comment]
Thanks. I think I'll come back with some more 64bit fixes if I'll get Chrome working with >4gb shared memory for pthreads. Right now i've managed to make it create WebAssembly.Memory but it just crashes after loading the code.

**Pull request merged**